drivers/adc/adc_mcp356x.c

```diff
@@ -0,0 +1,836 @@
+		{ .buf = &cmd, .len = sizeof(cmd) },
+	};
+	const struct spi_buf rx[] = {
+		{ .buf = &status, .len = sizeof(status) },
+		{ .buf = &data, .len = bytes },
+	};
+	const struct spi_buf_set tx_set = { .buffers = tx, .count = ARRAY_SIZE(tx) };
+	const struct spi_buf_set rx_set = { .buffers = rx, .count = ARRAY_SIZE(rx) };
+
+	int err = spi_transceive_dt(&config->bus, &tx_set, &rx_set);
+
+	if (bytes == 4) {
+		/* Buffer must be aligned */
+		*(uint32_t *)buffer = sys_be32_to_cpu(data);
+	} else {
+		/* Buffer might not be aligned */
+		sys_put_be24(data, buffer);
+	}
+	LOG_VDBG("D: res %d, status 0x%02x, %d 0x%0*x", err, status, bytes, bytes*2, data);
+
+	return err ? err : status;
+}
+
+
+static inline int reg_write(const struct device *dev, uint8_t reg, uint32_t val)
+{
+	return reg_write_size(dev, reg, reg_lens[reg], val, NULL);
+}
+
+/* Issue a "fast command", see datasheet §6.2.5. */
+static int fast_write(const struct device *dev, uint8_t fast_cmd)
+{
+	const struct mcp356x_config *config = dev->config;
+	uint8_t cmd = config->addr | fast_cmd | OP_FAST_CMD;
+	uint8_t status;
+	const struct spi_buf tx[] = {
+		{ .buf = &cmd, .len = sizeof(cmd) },
+	};
+	const struct spi_buf rx[] = {
+		{ .buf = &status, .len = sizeof(status) },
+	};
+	const struct spi_buf_set tx_set = { .buffers = tx, .count = ARRAY_SIZE(tx) };
+	const struct spi_buf_set rx_set = { .buffers = rx, .count = ARRAY_SIZE(rx) };
+
+	int err = spi_transceive_dt(&config->bus, &tx_set, &rx_set);
+
+	LOG_VDBG("%c: res %d, status 0x%02x, reg 0x%1x = 0x%0*x", 'C',
+		 err, status, 0, 2, cmd);
+
+	return err ? err : status;
+}
+
+static int get_bits_res(int resolution)
+{
+	switch (resolution) {
+	case 32:
+		return REG_CONFIG3_DATA_FORMAT_32_RJ;
+	case 24:
+		return REG_CONFIG3_DATA_FORMAT_24;
+	default:
+		return -EINVAL;
+	}
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct mcp356x_data *data = CONTAINER_OF(ctx, struct mcp356x_data, ctx);
+
+	if (repeat_sampling) {
+		/* I'm assuming this means start over from the beginning of the
+		 * buffer rather than repeat the last sample.
+		 */
+		data->write_ptr = ctx->sequence.buffer;
+	}
+	/* else, writeptr += num_ch, which was already done in the acq thread
+	 * after reading the samples, so we have nothing to do here.
+	 */
+}
+
+/* This kicks mcp356x_acquisition_thread to run */
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct mcp356x_data *data = CONTAINER_OF(ctx, struct mcp356x_data, ctx);
+
+	k_sem_give(&data->acq_sem);
+}
+
+/* Read all configured channels after each DRDY interrupt, signalled via &data->drdy_sem */
+static int sample_channels(const struct device *dev, uint32_t channels, int bps)
+{
+	struct mcp356x_data *data = dev->data;
+	int channel;
+
+	for (; channels; channels &= ~BIT(channel)) {
+		channel = find_msb_set(channels) - 1;
+
+		/* Wait for IRQ */
+		k_sem_take(&data->drdy_sem, K_SECONDS(12));
```

Comment:
I am sure the 12 seconds comes from somewhere, but please either use a define for it or document it.